src/server/coprocessor/executor/aggregation.rs

```diff
@@ -0,0 +1,273 @@
+    use util::codec::datum::{Datum, DatumDecoder};
+    use util::codec::number::NumberEncoder;
+    use util::codec::mysql::decimal::Decimal;
+    use util::codec::mysql::types;
+    use storage::Statistics;
+
+    use tipb::executor::TableScan;
+    use tipb::expression::{Expr, ExprType};
+
+    use protobuf::RepeatedField;
+
+    #[inline]
+    fn build_expr(tp: ExprType, id: Option<i64>, child: Option<Expr>) -> Expr {
+        let mut expr = Expr::new();
+        expr.set_tp(tp);
+        if tp == ExprType::ColumnRef {
+            expr.mut_val().encode_i64(id.unwrap()).unwrap();
+        } else {
+            expr.mut_children().push(child.unwrap());
+        }
+        expr
+    }
+
+    fn build_group_by(col_ids: &[i64]) -> Vec<Expr> {
+        let mut group_by = Vec::with_capacity(col_ids.len());
+        for id in col_ids {
+            group_by.push(build_expr(ExprType::ColumnRef, Some(*id), None));
+        }
+        group_by
+    }
+
+    fn build_aggr_func(aggrs: &[(ExprType, i64)]) -> Vec<Expr> {
+        let mut aggr_func = Vec::with_capacity(aggrs.len());
+        for aggr in aggrs {
+            let &(tp, id) = aggr;
+            let col_ref = build_expr(ExprType::ColumnRef, Some(id), None);
+            aggr_func.push(build_expr(tp, None, Some(col_ref)));
+        }
+        aggr_func
+    }
+
+    #[test]
+    fn test_aggregation() {
+        // prepare data and store
+        let tid = 1;
+        let cis = vec![new_col_info(1, types::LONG_LONG),
+                       new_col_info(2, types::VARCHAR),
+                       new_col_info(3, types::NEW_DECIMAL)];
+        let raw_data = vec![vec![Datum::I64(1), Datum::Bytes(b"a".to_vec()), Datum::Dec(7.into())],
+                            vec![Datum::I64(2), Datum::Bytes(b"a".to_vec()), Datum::Dec(7.into())],
+                            vec![Datum::I64(3), Datum::Bytes(b"b".to_vec()), Datum::Dec(8.into())],
+                            vec![Datum::I64(4), Datum::Bytes(b"a".to_vec()), Datum::Dec(7.into())],
+                            vec![Datum::I64(5), Datum::Bytes(b"f".to_vec()), Datum::Dec(5.into())],
+                            vec![Datum::I64(6), Datum::Bytes(b"b".to_vec()), Datum::Dec(8.into())],
+                            vec![Datum::I64(7), Datum::Bytes(b"f".to_vec()), Datum::Dec(6.into())]];
+        let table_data = gen_table_data(tid, &cis, &raw_data);
+        let mut test_store = TestStore::new(&table_data);
+        // init table scan meta
+        let mut table_scan = TableScan::new();
+        table_scan.set_table_id(tid);
+        table_scan.set_columns(RepeatedField::from_vec(cis.clone()));
+        // init TableScan Exectutor
+        let key_ranges = vec![get_range(tid, i64::MIN, i64::MAX)];
+        let (snapshot, start_ts) = test_store.get_snapshot();
+        let mut statistics = Statistics::default();
+        let mut ts_ect =
+            TableScanExecutor::new(table_scan, key_ranges, snapshot, &mut statistics, start_ts);
```

Comment:
Please fix the test, since we have added a new param for `TableScanExecutor::new`:
```
impl<'a> TableScanExecutor<'a> {
    #[allow(dead_code)] //TODO:remove it
    pub fn new(meta: TableScan,
               key_ranges: Vec<KeyRange>,
               snapshot: &'a Snapshot,
               statistics: &'a mut Statistics,
               start_ts: u64,
               isolation_level: IsolationLevel)
               -> TableScanExecutor<'a> 
```